diesel/src/sqlite/connection/raw.rs

```diff
@@ -292,99 +313,160 @@ extern "C" fn run_aggregator_step_function<ArgsSqlType, RetSqlType, Args, Ret, A
+    let result = args
+        .map(|args| std::panic::catch_unwind(move || Ok(aggregator.step(args))))
+        .unwrap_or_else(|e| Ok(Err(e)));
+    match result {
+        Ok(Ok(())) => (),
+        Ok(Err(e)) => {
+            let msg = e.to_string();
+            unsafe { context_error_str(ctx, &msg) };
+        }
+        Err(_) => {
+            let msg = format!("{}::step() panicked", std::any::type_name::<A>());
+            unsafe { context_error_str(ctx, &msg) };
+        }
+    };
 }
 
 extern "C" fn run_aggregator_final_function<ArgsSqlType, RetSqlType, Args, Ret, A>(
     ctx: *mut ffi::sqlite3_context,
 ) where
-    A: SqliteAggregateFunction<Args, Output = Ret> + 'static + Send,
+    A: SqliteAggregateFunction<Args, Output = Ret> + 'static + Send + std::panic::UnwindSafe,
     Args: FromSqlRow<ArgsSqlType, Sqlite>,
     Ret: ToSql<RetSqlType, Sqlite>,
     Sqlite: HasSqlType<RetSqlType>,
 {
-    unsafe {
+    let aggregate_context = unsafe {
         // Within the xFinal callback, it is customary to set nBytes to 0 so no pointless memory
         // allocations occur, a null pointer is returned in this case
         // See: https://www.sqlite.org/c3ref/aggregate_context.html
         //
         // For the reasoning about the safety of the OptionalAggregator handling
         // see the comment in run_aggregator_step_function.
-        let aggregate_context = ffi::sqlite3_aggregate_context(ctx, 0);
-        let mut aggregate_context = NonNull::new(aggregate_context as *mut OptionalAggregator<A>);
-        let aggregator = match aggregate_context {
-            Some(ref mut a) => match std::mem::replace(a.as_mut(), OptionalAggregator::None) {
-                OptionalAggregator::Some(agg) => Some(agg),
-                OptionalAggregator::None => unreachable!("We've written to the aggregator in the xStep callback. If xStep was never called, then ffi::sqlite_aggregate_context() would have returned a NULL pointer")
-            },
-            None => None,
-        };
-
-        let result = A::finalize(aggregator);
+        ffi::sqlite3_aggregate_context(ctx, 0)
+    };
 
-        match process_sql_function_result::<RetSqlType, Ret>(result) {
-            Ok(value) => value.result_of(ctx),
-            Err(e) => {
-                let msg = e.to_string();
-                ffi::sqlite3_result_error(ctx, msg.as_ptr() as *const _, msg.len() as _);
+    let mut aggregate_context = NonNull::new(aggregate_context as *mut OptionalAggregator<A>);
+    let aggregator = aggregate_context.as_mut().map(|a| {
+        let a = unsafe { a.as_mut() };
+        match std::mem::replace(a, OptionalAggregator::None) {
+            OptionalAggregator::Some(agg) => agg,
+            OptionalAggregator::None => {
+                eprintln!("We've written to the aggregator in the xStep callback. If xStep was never called, then ffi::sqlite_aggregate_context() would have returned a NULL pointer");
```

Comment:
Could you improve that error message saying something like: If you ever see this message, please open an issue at the diesel github repository?


(Maybe with an indication on which line this error happened?)